### PR TITLE
RD-1107/multi-process-gap-detector

### PR DESCRIPTION
**Pull request by BlayneJLees**
# RD-1107/multi-process-gap-detector #

## Description ##

This PR multi-processes the gap detector and also introduces the argument to run with curved rows.

Jira ticket : [RD-1107](https://hummingbirdtech.atlassian.net/browse/RD-1107)

## How To Test ##

This has been tested with surveys live on the platform.

## Checklist ##

- [x] PEP8 compliant (run `black --check $(git diff --name-only HEAD origin/develop | grep .py) -l 119`);
- [ ] Passing tests;
- [ ] _(If required)_ Confluence pages updated.

### CODE REVIEW

File: hummingbirdtech/planting_lines/__init__.py [hunk 1]
```diff
@@ -5,7 +5,7 @@
 from .line_reconstruction.dic import parameters as line_reconstruction_parameters
 
 
-__version__ = "1.1.1"
+__version__ = "1.1.2"
```
Comment (bszadman) [inline hummingbirdtech/planting_lines/__init__.py]: 1.2.0 I woud say, as you are implementing new functionalities

Comment (BlayneJLees) [reply hummingbirdtech/planting_lines/__init__.py]: Ok yes good point, alot has changed here!

File: setup.py [hunk 2]
```diff
@@ -53,8 +53,8 @@
     ],
     entry_points={
         'console_scripts': [
-            'hb-gap-detection=hummingbirdtech.planting_lines.gap_detection.cli:main',
-            'hb-line-reconstruction=hummingbirdtech.planting_lines.line_reconstruction.cli:main',
+            'hb-gap_detection=hummingbirdtech.planting_lines.gap_detection.cli:main',
```
Comment (vgcornwell) [inline setup.py]: Why have these changed to `_` now?

Comment (BlayneJLees) [reply setup.py]: ahhh very nice I think its how ive split the prs somehow!

File: hummingbirdtech/planting_lines/gap_detection/algorithm/gap_detector.py [hunk 3]
```diff
@@ -56,40 +61,60 @@ def get_gap_lines(self, vegetation_band, crop_row_lines):
...
+        return list(
+            self.flatten(
+                self._get_line_list(
+                    crop_row_lines=crop_row_lines,
+                    vegetation_polygons=flattened_vegetation_polygons,
+                )
+            )
+        )
+
+    def _get_line_list(self, crop_row_lines, vegetation_polygons):
```
Comment (aj-hitchins) [inline hummingbirdtech/planting_lines/gap_detection/algorithm/gap_detector.py]: Docstrings.

File: hummingbirdtech/planting_lines/gap_detection/algorithm/gap_detector.py [hunk 4]
```diff
@@ -56,40 +61,60 @@ def get_gap_lines(self, vegetation_band, crop_row_lines):
...
+        for multi_linestring in filtered_lines:
             if type(multi_linestring) == MultiLineString:
-                line_list.append([line for line in multi_linestring])
+                yield [line for line in multi_linestring]
             else:
-                line_list.append(multi_linestring)
-        return list(self.flatten(line_list))
+                yield multi_linestring
+
+    def _get_difference_line(self, line, vegetation_polygons):
```
Comment (aj-hitchins) [inline hummingbirdtech/planting_lines/gap_detection/algorithm/gap_detector.py]: Docstrings

File: hummingbirdtech/planting_lines/gap_detection/algorithm/gap_detector.py [hunk 5]
```diff
@@ -56,40 +61,60 @@ def get_gap_lines(self, vegetation_band, crop_row_lines):
...
+        candidate_polygons = self._get_candidate_polygons(vegetation_polygons, line)
+        # Buffer used to overcome invalid geometry error for intersecting polygons
+        multipolygon = MultiPolygon(candidate_polygons).buffer(0)
+        return line.difference(multipolygon)
+
+    def _get_candidate_polygons(self, vegetation_polygons, line):
```
Comment (aj-hitchins) [inline hummingbirdtech/planting_lines/gap_detection/algorithm/gap_detector.py]: Docstrings

File: hummingbirdtech/planting_lines/gap_detection/algorithm/parallel_map.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
+from multiprocessing import Queue, Process
+
+
+# f is any function that takes an item as part of an iterable and then does some actual processing.
+
+
+def manager(f, queue_in, queue_out, kwargs):
```
Comment (aj-hitchins) [inline hummingbirdtech/planting_lines/gap_detection/algorithm/parallel_map.py]: Docstrings

File: hummingbirdtech/planting_lines/gap_detection/algorithm/parallel_map.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    """Takes items from a queue (populated from the iterable), and puts the function output into another queue.
+    """
+    while True:
+        index, item = queue_in.get()
+        if index is None:
+            break
+        queue_out.put((index, f(item, **kwargs)))
+
+
+def parmap(f, iterator, n_processes, **kwargs):
```
Comment (aj-hitchins) [inline hummingbirdtech/planting_lines/gap_detection/algorithm/parallel_map.py]: Docstrings

File: hummingbirdtech/planting_lines/gap_detection/algorithm/gap_detector.py [hunk 8]
```diff
@@ -31,6 +35,7 @@ def __init__(
         self._contour_value = contour_value
         self._sigma = sigma
         self._logger = logger
+        self._cpu_count = cpu_count() - 2
```
[the remark this answers is not in the document]
Comment (BlayneJLees) [reply hummingbirdtech/planting_lines/gap_detection/algorithm/gap_detector.py]: I spoke to aj and Austin and actually they said its fine for it to be the `cpu_count`, which i'll change it to. I'll make a note that if this is being run locally perhaps there is the risk that your computer could crash and to reduce this number. However, for the vms this shouldn't be an issue!

File: hummingbirdtech/planting_lines/gap_detection/algorithm/parallel_map.py (hunk 7 above)
Comment (BlayneJLees) [reply hummingbirdtech/planting_lines/gap_detection/algorithm/parallel_map.py]: Cheers!

File: hummingbirdtech/planting_lines/gap_detection/algorithm/parallel_map.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
+from multiprocessing import Queue, Process
+
+
+# f is any function that takes an item as part of an iterable and then does some actual processing.
+def manager(f, queue_in, queue_out, kwargs):
+    """Takes items from a queue (populated from the iterable), and puts the function output into another queue.
+    :type queue_in: multiprocessing.queues.Queue
```
Comment (aj-hitchins) [inline hummingbirdtech/planting_lines/gap_detection/algorithm/parallel_map.py]: missing `f`
```
:type f: callable
```

**Pull request merged**